File: core/src/nl/t64/game/rpg/screens/world/GameMapQuestDiscover.java
package nl.t64.game.rpg.screens.world;

import com.badlogic.gdx.maps.MapObject;
import com.badlogic.gdx.maps.objects.RectangleMapObject;
import com.badlogic.gdx.math.Rectangle;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.components.quest.QuestGraph;


class GameMapQuestDiscover {

    final Rectangle rectangle;
    final String questId;
    final String taskId;

    GameMapQuestDiscover(MapObject mapObject) {
        RectangleMapObject rectObject = (RectangleMapObject) mapObject;

        this.rectangle = rectObject.getRectangle();
        this.questId = rectObject.getProperties().get("type", String.class);
        this.taskId = rectObject.getProperties().get("task", String.class);
    }

    void locationDiscovered() {
        QuestGraph quest = Utils.getGameData().getQuests().getQuestById(questId);
        quest.setLocationDiscoverd(taskId);
    }

}

File: core/src/nl/t64/game/rpg/screens/world/mapobjects/GameMapQuestDiscover.kt
package nl.t64.game.rpg.screens.world.mapobjects

import com.badlogic.gdx.maps.objects.RectangleMapObject
import com.badlogic.gdx.math.Rectangle
import ktx.tiled.property
import nl.t64.game.rpg.Utils.brokerManager
import nl.t64.game.rpg.Utils.gameData
import nl.t64.game.rpg.screens.world.entity.Direction
import nl.t64.game.rpg.subjects.CollisionObserver


class GameMapQuestDiscover(rectObject: RectangleMapObject) : GameMapObject(rectObject.rectangle), CollisionObserver {

    private val questId: String = rectObject.name
    private val taskId: String = rectObject.property("task")

    init {
        brokerManager.collisionObservers.addObserver(this)
    }

    override fun onNotifyCollision(playerBoundingBox: Rectangle, playerDirection: Direction) {
        if (playerBoundingBox.overlaps(rectangle)) {
            setQuestTaskComplete()
        }
    }

    private fun setQuestTaskComplete() {
        val quest = gameData.quests.getQuestById(questId)
        quest.setTaskComplete(taskId)
    }

}
